File: intern/647.palindromic-substrings.js
/*
 * @lc app=leetcode id=647 lang=javascript
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
const countSubstrings = function (s) {
    const LENGTH = s.length;
    /**
     * 从 s[i] 到 s[j] 是否是回文
     * @type {boolean[][]} */
    const dp = new Array(LENGTH);
    for (let i = 0; i < LENGTH; i++) {
        dp[i] = new Array(LENGTH);
        dp[i].fill(false);
    }

    let count = 0;
    for (let i = LENGTH - 1; i >= 0; i--) {
        for (let j = i; j < LENGTH; j++) {
            if (s[i] === s[j]) {
                if (i === j) {
                    count++;
                    dp[i][j] = true;
                } else {
                    if (j === i + 1) {
                        count++;
                        dp[i][j] = true;
                    }
                    if (j === i + 2) {
                        count++;
                        dp[i][j] = true;
                    } else if (dp[i + 1][j - 1] === true) {
                        count++;
                        dp[i][j] = true;
                    }
                }
            }
        }
    }
    return count;
};
// @lc code=end

console.log(countSubstrings('aaaaa'));

File: prev-2024/647.palindromic-substrings.cpp
/*
 * @lc app=leetcode id=647 lang=cpp
 *
 * [647] Palindromic Substrings
 */
#include <algorithm>
#include <string>
#include <vector>
using std::string;
using std::vector;

// @lc code=start
/**
 * isPalindrome[i][j]: s[i] - s[j] is palindrome
 *
 * base case
 * isPalindrome[i][i] = true
 *
 * isPalindrome[i][j] = s[i] == s[j] && isPalindrome[i+1][j-1]
 */
class Solution {
 public:
  int countSubstrings(string s) {
    int palindromeSubstringCount = s.size();

    bool** isPalindrome = new bool*[2];
    for (int i = 0; i < 2; i++) {
      isPalindrome[i] = new bool[s.size()];
    }

    for (int i = s.size() - 1; i >= 0; i--) {
      isPalindrome[i % 2][i] = true;

      for (int j = i + 1; j < s.size(); j++) {
        if (i + 1 == j) {
          isPalindrome[i % 2][j] = s[i] == s[j];
        } else {
          isPalindrome[i % 2][j] =
              s[i] == s[j] && isPalindrome[(i + 1) % 2][j - 1];
        }

        if (isPalindrome[i % 2][j]) {
          palindromeSubstringCount++;
        }
      }
    }

    for (int i = 0; i < 2; i++) {
      delete[] isPalindrome[i];
    }
    delete[] isPalindrome;

    return palindromeSubstringCount;
  }
};
// @lc code=end
